File: packages/api/src/lib/databases-config.ts
import * as os from "os";
import * as path from "path";
import { readJson, readToml, writeToml, pathExists, remove } from "node-utils";
//
// A database entry stored in databases.toml.
// The name field is the unique (case-insensitive) identifier for each entry.
//
export interface IDatabaseEntry {
    // Human-readable display name.
    name: string;

    // Optional description of this database.
    description: string;

    // Absolute filesystem path (or S3 path) to the database directory.
    path: string;

    // Optional origin string read from .db/config.json; refreshed each time the database is opened.
    origin?: string;

    // Vault secret name for S3 credentials.
    s3Key?: string;

    // Vault secret name for the encryption key pair.
    encryptionKey?: string;

    // Vault secret name for the geocoding API key.
    geocodingKey?: string;
}

//
// Configuration for the databases list, stored in ~/.config/photosphere/databases.toml.
//
interface IDatabasesConfig {
    //
    // Structured list of configured databases.
    //
    databases: IDatabaseEntry[];

    //
    // Ordered list of recently opened database names (most recent first, max 5).
    //
    recentDatabaseNames: string[];
}

//
// TOML on-disk shape for a single database entry (snake_case keys).
//
interface ITomlDatabaseEntry {
    // Human-readable display name.
    name: string;

    // Optional description of this database.
    description: string;

    // Absolute filesystem path (or S3 path) to the database directory.
    path: string;

    // Optional origin string.
    origin?: string;

    // Vault secret name for S3 credentials.
    s3_key?: string;

    // Vault secret name for the encryption key pair.
    encryption_key?: string;

    // Vault secret name for the geocoding API key.
    geocoding_key?: string;
}

//
// TOML on-disk shape for the databases config file (snake_case keys).
//
interface ITomlDatabasesConfig {
    // Array of database entries.
    databases?: ITomlDatabaseEntry[];

    // Recently opened database names.
    recent_database_names?: string[];

    // Legacy field — recently opened database paths. Migrated on load.
    recent_database_paths?: string[];
}

const CONFIG_DIR = process.env.PHOTOSPHERE_CONFIG_DIR || path.join(os.homedir(), ".config", "photosphere");
const DATABASES_FILE = path.join(CONFIG_DIR, "databases.toml");
const OLD_DATABASES_FILE = path.join(CONFIG_DIR, "databases.json");

//
// Converts a TOML-shaped database entry to the TypeScript IDatabaseEntry type.
//
function tomlEntryToDatabaseEntry(tomlEntry: ITomlDatabaseEntry): IDatabaseEntry {
    const entry: IDatabaseEntry = {
        name: tomlEntry.name,
        description: tomlEntry.description,
        path: tomlEntry.path,
    };
    if (tomlEntry.origin !== undefined) {
        entry.origin = tomlEntry.origin;
    }
    if (tomlEntry.s3_key !== undefined) {
        entry.s3Key = tomlEntry.s3_key;
    }
    if (tomlEntry.encryption_key !== undefined) {
        entry.encryptionKey = tomlEntry.encryption_key;
    }
    if (tomlEntry.geocoding_key !== undefined) {
        entry.geocodingKey = tomlEntry.geocoding_key;
    }
    return entry;
}

//
// Converts a TypeScript IDatabaseEntry to the TOML on-disk shape.
//
function databaseEntryToToml(entry: IDatabaseEntry): ITomlDatabaseEntry {
    const tomlEntry: ITomlDatabaseEntry = {
        name: entry.name,
        description: entry.description,
        path: entry.path,
    };
    if (entry.origin !== undefined) {
        tomlEntry.origin = entry.origin;
    }
    if (entry.s3Key !== undefined) {
        tomlEntry.s3_key = entry.s3Key;
    }
    if (entry.encryptionKey !== undefined) {
        tomlEntry.encryption_key = entry.encryptionKey;
    }
    if (entry.geocodingKey !== undefined) {
        tomlEntry.geocoding_key = entry.geocodingKey;
    }
    return tomlEntry;
}

//
// Converts a TOML-shaped config object to the TypeScript IDatabasesConfig type.
// Recognises only the new `recent_database_names` field; legacy `recent_database_paths`
// migration is handled separately in `loadDatabasesConfig`.
//
function tomlToDatabasesConfig(toml: ITomlDatabasesConfig): IDatabasesConfig {
    const databases = Array.isArray(toml.databases)
        ? toml.databases.map(tomlEntryToDatabaseEntry)
        : [];
    const recentDatabaseNames = Array.isArray(toml.recent_database_names)
        ? toml.recent_database_names
        : [];
    return { databases, recentDatabaseNames };
}

//
// Converts the TypeScript IDatabasesConfig to the TOML on-disk shape.
//
function databasesConfigToToml(config: IDatabasesConfig): ITomlDatabasesConfig {
    return {
        databases: config.databases.map(databaseEntryToToml),
        recent_database_names: config.recentDatabaseNames,
    };
}

//
// Returns true if the two names match case-insensitively.
//
function namesMatch(left: string, right: string): boolean {
    return left.toLowerCase() === right.toLowerCase();
}

//
// Loads the databases configuration from disk.
// If the TOML file does not exist but an old JSON file does, migrates automatically.
// If the loaded TOML still uses the legacy `recent_database_paths` field, converts it
// to `recent_database_names` (resolving each path to its current entry's name; dropping
// paths that no longer match any entry) and rewrites the file.
// Returns a default config with an empty list if neither file exists.
//
export async function loadDatabasesConfig(): Promise<IDatabasesConfig> {
    if (!await pathExists(DATABASES_FILE)) {
        if (await pathExists(OLD_DATABASES_FILE)) {
            const jsonConfig = await readJson<{ databases?: IDatabaseEntry[]; recentDatabasePaths?: string[]; recentDatabaseNames?: string[] }>(OLD_DATABASES_FILE);
            const databases = Array.isArray(jsonConfig.databases) ? jsonConfig.databases : [];
            let recentDatabaseNames: string[];
            if (Array.isArray(jsonConfig.recentDatabaseNames)) {
                recentDatabaseNames = jsonConfig.recentDatabaseNames;
            }
            else if (Array.isArray(jsonConfig.recentDatabasePaths)) {
                recentDatabaseNames = recentPathsToNames(jsonConfig.recentDatabasePaths, databases);
            }
            else {
                recentDatabaseNames = [];
            }
            const migrated: IDatabasesConfig = { databases, recentDatabaseNames };
            await saveDatabasesConfig(migrated);
            await remove(OLD_DATABASES_FILE);
            return migrated;
        }
        return { databases: [], recentDatabaseNames: [] };
    }

    const toml = await readToml<ITomlDatabasesConfig>(DATABASES_FILE);

    // Legacy migration: convert recent_database_paths to recent_database_names and rewrite the file once.
    if (!Array.isArray(toml.recent_database_names) && Array.isArray(toml.recent_database_paths)) {
        const databases = Array.isArray(toml.databases)
            ? toml.databases.map(tomlEntryToDatabaseEntry)
            : [];
        const migrated: IDatabasesConfig = {
            databases,
            recentDatabaseNames: recentPathsToNames(toml.recent_database_paths, databases),
        };
        await saveDatabasesConfig(migrated);
        return migrated;
    }

    return tomlToDatabasesConfig(toml);
}

//
// Resolves the legacy recent-paths array into the new recent-names array by looking up
// each path in the current databases list. Paths that don't match any entry are dropped.
//
function recentPathsToNames(recentPaths: string[], databases: IDatabaseEntry[]): string[] {
    const result: string[] = [];
    for (const recentPath of recentPaths) {
        const match = databases.find(dbEntry => dbEntry.path === recentPath);
        if (match) {
            result.push(match.name);
        }
    }
    return result;
}

//
// Saves the databases configuration to disk.
//
export async function saveDatabasesConfig(config: IDatabasesConfig): Promise<void> {
    if (!Array.isArray(config.databases)) {
        config.databases = [];
    }
    if (!Array.isArray(config.recentDatabaseNames)) {
        config.recentDatabaseNames = [];
    }
    await writeToml(DATABASES_FILE, databasesConfigToToml(config));
}

//
// Returns all configured database entries.
//
export async function getDatabases(): Promise<IDatabaseEntry[]> {
    const config = await loadDatabasesConfig();
    return config.databases;
}

//
// Finds a database entry by name using case-insensitive matching.
// Returns the first match if any. Returns undefined if no entry matches.
//
export async function findDatabase(name: string): Promise<IDatabaseEntry | undefined> {
    const config = await loadDatabasesConfig();
    return config.databases.find(dbEntry => namesMatch(dbEntry.name, name));
}

//
// Adds a new database entry to the list.
// Throws if an entry with the same name (case-insensitive) already exists; this acts as
// a storage-layer invariant in addition to any UX checks.
//
export async function addDatabaseEntry(entry: IDatabaseEntry): Promise<void> {
    const config = await loadDatabasesConfig();
    const existing = config.databases.find(dbEntry => namesMatch(dbEntry.name, entry.name));
    if (existing) {
        throw new Error(`A database named "${entry.name}" already exists.`);
    }
    config.databases = [...config.databases, entry];
    await saveDatabasesConfig(config);
}

//
// Updates the entry currently identified by `originalName` with the new fields in `entry`.
// If the new entry's name differs from `originalName`, the matching slot in
// `recentDatabaseNames` is rewritten to keep the recents list pointing at the same entry.
// Throws if the rename would collide with another existing entry, or if no entry with
// `originalName` is found.
//
export async function updateDatabaseEntry(originalName: string, entry: IDatabaseEntry): Promise<void> {
    const config = await loadDatabasesConfig();
    const matchIndex = config.databases.findIndex(dbEntry => namesMatch(dbEntry.name, originalName));
    if (matchIndex === -1) {
        throw new Error(`No database named "${originalName}" found.`);
    }
    const renamed = !namesMatch(entry.name, originalName);
    if (renamed) {
        const collision = config.databases.find((dbEntry, dbIndex) => dbIndex !== matchIndex && namesMatch(dbEntry.name, entry.name));
        if (collision) {
            throw new Error(`A database named "${entry.name}" already exists.`);
        }
    }
    const updatedDatabases = config.databases.slice();
    updatedDatabases[matchIndex] = entry;
    config.databases = updatedDatabases;
    if (renamed) {
        config.recentDatabaseNames = config.recentDatabaseNames.map(recentName => namesMatch(recentName, originalName) ? entry.name : recentName);
    }
    await saveDatabasesConfig(config);
}

//
// Removes a database entry by name (case-insensitive).
// Removes only the first matching entry from `databases` (defensive against legacy state
// where two entries share a name). Also removes the same name from `recentDatabaseNames`.
// No-op if no entry matches.
//
export async function removeDatabaseEntry(name: string): Promise<void> {
    const config = await loadDatabasesConfig();
    const matchIndex = config.databases.findIndex(dbEntry => namesMatch(dbEntry.name, name));
    if (matchIndex === -1) {
        // Still clean recents in case of stale state.
        const filteredRecents = config.recentDatabaseNames.filter(recentName => !namesMatch(recentName, name));
        if (filteredRecents.length !== config.recentDatabaseNames.length) {
            config.recentDatabaseNames = filteredRecents;
            await saveDatabasesConfig(config);
        }
        return;
    }
    const updatedDatabases = config.databases.slice();
    updatedDatabases.splice(matchIndex, 1);
    config.databases = updatedDatabases;
    config.recentDatabaseNames = config.recentDatabaseNames.filter(recentName => !namesMatch(recentName, name));
    await saveDatabasesConfig(config);
}

//
// Returns the top-5 most recently opened databases, ordered most-recent first.
// Names that no longer resolve to an entry in the databases list are silently dropped.
//
export async function getRecentDatabases(): Promise<IDatabaseEntry[]> {
    const config = await loadDatabasesConfig();
    const result: IDatabaseEntry[] = [];
    for (const recentName of config.recentDatabaseNames) {
        const found = config.databases.find(dbEntry => namesMatch(dbEntry.name, recentName));
        if (found) {
            result.push(found);
        }
    }
    return result;
}

//
// Removes the given name from recentDatabaseNames only. Leaves the matching entry
// in `databases` untouched. No-op if the name is not in the recent list.
//
export async function removeRecentDatabaseName(name: string): Promise<void> {
    const config = await loadDatabasesConfig();
    const filtered = config.recentDatabaseNames.filter(recentName => !namesMatch(recentName, name));
    if (filtered.length === config.recentDatabaseNames.length) {
        return;
    }
    config.recentDatabaseNames = filtered;
    await saveDatabasesConfig(config);
}

//
// Moves the database entry matching the given name (case-insensitive) to the front of
// recentDatabaseNames, trimming the list to a maximum of 5 entries, then saves.
// No-op if no entry matches.
//
export async function markDatabaseOpened(name: string): Promise<void> {
    const config = await loadDatabasesConfig();
    const found = config.databases.find(dbEntry => namesMatch(dbEntry.name, name));
    if (!found) {
        return;
    }
    config.recentDatabaseNames = [
        found.name,
        ...config.recentDatabaseNames.filter(recentName => !namesMatch(recentName, found.name)),
    ].slice(0, 5);
    await saveDatabasesConfig(config);
}

File: packages/node-api-zig/src/lib/databases-config.zig
const std = @import("std");
const builtin = @import("builtin");
const node_utils = @import("node-utils-zig");
const fs = node_utils.fs;
const process_env = node_utils.process_env;

//
// A database entry stored in databases.toml.
// The name field is the unique (case-insensitive) identifier for each entry.
//
pub const IDatabaseEntry = struct {
    // Human-readable display name.
    name: []const u8,

    // Optional description of this database.
    description: []const u8,

    // Absolute filesystem path (or S3 path) to the database directory.
    path: []const u8,

    // Optional origin string read from .db/config.json; refreshed each time the database is opened.
    origin: ?[]const u8 = null,

    // Vault secret name for S3 credentials.
    s3Key: ?[]const u8 = null,

    // Vault secret name for the encryption key pair.
    encryptionKey: ?[]const u8 = null,

    // Vault secret name for the geocoding API key.
    geocodingKey: ?[]const u8 = null,
};

//
// Configuration for the databases list, stored in ~/.config/photosphere/databases.toml.
//
pub const IDatabasesConfig = struct {
    //
    // Structured list of configured databases.
    //
    databases: []const IDatabaseEntry,

    //
    // Ordered list of recently opened database names (most recent first, max 5).
    //
    recentDatabaseNames: []const []const u8,
};

//
// (Zig: the TOML on-disk shapes ITomlDatabaseEntry and ITomlDatabasesConfig (snake_case keys) are the std.json.Value
// objects produced by node-utils-zig's TOML parser, so they have no separate struct types.)
//

//
// The directory holding the databases config.
// (TypeScript: the module constant CONFIG_DIR, which reads the environment when the module loads; Zig reads it on
// each call.)
//
fn CONFIG_DIR(allocator: std.mem.Allocator) ![]const u8 {
    if (process_env.getEnv("PHOTOSPHERE_CONFIG_DIR")) |configDir| {
        if (configDir.len > 0) {
            return configDir;
        }
    }
    const homeDir = process_env.getEnv(if (builtin.os.tag == .windows) "USERPROFILE" else "HOME") orelse "";
    return std.fs.path.join(allocator, &.{ homeDir, ".config", "photosphere" });
}

//
// The path of databases.toml (TypeScript: the module constant DATABASES_FILE).
//
fn DATABASES_FILE(allocator: std.mem.Allocator) ![]const u8 {
    return std.fs.path.join(allocator, &.{ try CONFIG_DIR(allocator), "databases.toml" });
}

//
// The path of the legacy databases.json (TypeScript: the module constant OLD_DATABASES_FILE).
//
fn OLD_DATABASES_FILE(allocator: std.mem.Allocator) ![]const u8 {
    return std.fs.path.join(allocator, &.{ try CONFIG_DIR(allocator), "databases.json" });
}

//
// Gets a string property of a JSON or TOML object (null when it is absent or not a string).
// (No TypeScript counterpart: TypeScript reads the property directly.)
//
fn stringProperty(object: std.json.ObjectMap, key: []const u8) ?[]const u8 {
    const value = object.get(key) orelse {
        return null;
    };
    return switch (value) {
        .string => |text| text,
        else => null,
    };
}

//
// Gets an array property of a JSON or TOML object (null when it is absent or not an array, like Array.isArray).
// (No TypeScript counterpart.)
//
fn arrayProperty(object: std.json.ObjectMap, key: []const u8) ?[]const std.json.Value {
    const value = object.get(key) orelse {
        return null;
    };
    return switch (value) {
        .array => |array| array.items,
        else => null,
    };
}

//
// Gets the strings of an array of strings (non-string items are skipped). (No TypeScript counterpart.)
//
fn stringItems(allocator: std.mem.Allocator, items: []const std.json.Value) ![]const []const u8 {
    var strings: std.ArrayList([]const u8) = .empty;
    for (items) |item| {
        switch (item) {
            .string => |text| try strings.append(allocator, text),
            else => {},
        }
    }
    return strings.items;
}

//
// Converts a TOML-shaped database entry to the TypeScript IDatabaseEntry type.
// (Zig: a missing or non-string name, description or path reads as "".)
//
fn tomlEntryToDatabaseEntry(tomlEntry: std.json.Value) IDatabaseEntry {
    const object = switch (tomlEntry) {
        .object => |object| object,
        else => std.json.ObjectMap.empty,
    };
    var entry: IDatabaseEntry = .{
        .name = stringProperty(object, "name") orelse "",
        .description = stringProperty(object, "description") orelse "",
        .path = stringProperty(object, "path") orelse "",
    };
    if (stringProperty(object, "origin")) |origin| {
        entry.origin = origin;
    }
    if (stringProperty(object, "s3_key")) |s3Key| {
        entry.s3Key = s3Key;
    }
    if (stringProperty(object, "encryption_key")) |encryptionKey| {
        entry.encryptionKey = encryptionKey;
    }
    if (stringProperty(object, "geocoding_key")) |geocodingKey| {
        entry.geocodingKey = geocodingKey;
    }
    return entry;
}

//
// Converts a JSON database entry of the legacy databases.json (camelCase keys) to IDatabaseEntry.
// (No TypeScript counterpart: TypeScript uses the parsed JSON objects as they are.)
//
fn jsonEntryToDatabaseEntry(jsonEntry: std.json.Value) IDatabaseEntry {
    const object = switch (jsonEntry) {
        .object => |object| object,
        else => std.json.ObjectMap.empty,
    };
    return .{
        .name = stringProperty(object, "name") orelse "",
        .description = stringProperty(object, "description") orelse "",
        .path = stringProperty(object, "path") orelse "",
        .origin = stringProperty(object, "origin"),
        .s3Key = stringProperty(object, "s3Key"),
        .encryptionKey = stringProperty(object, "encryptionKey"),
        .geocodingKey = stringProperty(object, "geocodingKey"),
    };
}

//
// Converts a TypeScript IDatabaseEntry to the TOML on-disk shape.
//
fn databaseEntryToToml(allocator: std.mem.Allocator, entry: IDatabaseEntry) !std.json.Value {
    var tomlEntry: std.json.ObjectMap = .empty;
    try tomlEntry.put(allocator, "name", .{ .string = entry.name });
    try tomlEntry.put(allocator, "description", .{ .string = entry.description });
    try tomlEntry.put(allocator, "path", .{ .string = entry.path });
    if (entry.origin) |origin| {
        try tomlEntry.put(allocator, "origin", .{ .string = origin });
    }
    if (entry.s3Key) |s3Key| {
        try tomlEntry.put(allocator, "s3_key", .{ .string = s3Key });
    }
    if (entry.encryptionKey) |encryptionKey| {
        try tomlEntry.put(allocator, "encryption_key", .{ .string = encryptionKey });
    }
    if (entry.geocodingKey) |geocodingKey| {
        try tomlEntry.put(allocator, "geocoding_key", .{ .string = geocodingKey });
    }
    return .{ .object = tomlEntry };
}

//
// Converts a TOML-shaped config object to the TypeScript IDatabasesConfig type.
// Recognises only the new `recent_database_names` field; legacy `recent_database_paths`
// migration is handled separately in `loadDatabasesConfig`.
//
fn tomlToDatabasesConfig(allocator: std.mem.Allocator, toml: std.json.ObjectMap) !IDatabasesConfig {
    var databases: std.ArrayList(IDatabaseEntry) = .empty;
    if (arrayProperty(toml, "databases")) |tomlDatabases| {
        for (tomlDatabases) |tomlEntry| {
            try databases.append(allocator, tomlEntryToDatabaseEntry(tomlEntry));
        }
    }
    const recentDatabaseNames = if (arrayProperty(toml, "recent_database_names")) |names|
        try stringItems(allocator, names)
    else
        &[_][]const u8{};
    return .{ .databases = databases.items, .recentDatabaseNames = recentDatabaseNames };
}

//
// Converts the TypeScript IDatabasesConfig to the TOML on-disk shape.
//
fn databasesConfigToToml(allocator: std.mem.Allocator, config: IDatabasesConfig) !std.json.Value {
    var databases = std.json.Array.init(allocator);
    for (config.databases) |entry| {
        try databases.append(try databaseEntryToToml(allocator, entry));
    }
    var recentDatabaseNames = std.json.Array.init(allocator);
    for (config.recentDatabaseNames) |recentName| {
        try recentDatabaseNames.append(.{ .string = recentName });
    }
    var toml: std.json.ObjectMap = .empty;
    try toml.put(allocator, "databases", .{ .array = databases });
    try toml.put(allocator, "recent_database_names", .{ .array = recentDatabaseNames });
    return .{ .object = toml };
}

// Not ported: namesMatch (only used by the functions below that are not ported).

//
// Loads the databases configuration from disk.
// If the TOML file does not exist but an old JSON file does, migrates automatically.
// If the loaded TOML still uses the legacy `recent_database_paths` field, converts it
// to `recent_database_names` (resolving each path to its current entry's name; dropping
// paths that no longer match any entry) and rewrites the file.
// Returns a default config with an empty list if neither file exists.
//
pub fn loadDatabasesConfig(allocator: std.mem.Allocator, io: std.Io) !IDatabasesConfig {
    const databasesFile = try DATABASES_FILE(allocator);
    const oldDatabasesFile = try OLD_DATABASES_FILE(allocator);
    if (!fs.pathExists(io, databasesFile)) {
        if (fs.pathExists(io, oldDatabasesFile)) {
            const jsonValue = try fs.readJson(allocator, io, oldDatabasesFile);
            const jsonConfig = switch (jsonValue) {
                .object => |object| object,
                else => std.json.ObjectMap.empty,
            };
            var databases: std.ArrayList(IDatabaseEntry) = .empty;
            if (arrayProperty(jsonConfig, "databases")) |jsonDatabases| {
                for (jsonDatabases) |jsonEntry| {
                    try databases.append(allocator, jsonEntryToDatabaseEntry(jsonEntry));
                }
            }
            var recentDatabaseNames: []const []const u8 = undefined;
            if (arrayProperty(jsonConfig, "recentDatabaseNames")) |names| {
                recentDatabaseNames = try stringItems(allocator, names);
            }
            else if (arrayProperty(jsonConfig, "recentDatabasePaths")) |paths| {
                recentDatabaseNames = try recentPathsToNames(allocator, try stringItems(allocator, paths), databases.items);
            }
            else {
                recentDatabaseNames = &.{};
            }
            const migrated: IDatabasesConfig = .{ .databases = databases.items, .recentDatabaseNames = recentDatabaseNames };
            try saveDatabasesConfig(allocator, io, migrated);
            try fs.remove(io, oldDatabasesFile);
            return migrated;
        }
        return .{ .databases = &.{}, .recentDatabaseNames = &.{} };
    }

    const tomlValue = try fs.readToml(allocator, io, databasesFile);
    const toml = switch (tomlValue) {
        .object => |object| object,
        else => std.json.ObjectMap.empty,
    };

    // Legacy migration: convert recent_database_paths to recent_database_names and rewrite the file once.
    if (arrayProperty(toml, "recent_database_names") == null and arrayProperty(toml, "recent_database_paths") != null) {
        var databases: std.ArrayList(IDatabaseEntry) = .empty;
        if (arrayProperty(toml, "databases")) |tomlDatabases| {
            for (tomlDatabases) |tomlEntry| {
                try databases.append(allocator, tomlEntryToDatabaseEntry(tomlEntry));
            }
        }
        const migrated: IDatabasesConfig = .{
            .databases = databases.items,
            .recentDatabaseNames = try recentPathsToNames(allocator, try stringItems(allocator, arrayProperty(toml, "recent_database_paths").?), databases.items),
        };
        try saveDatabasesConfig(allocator, io, migrated);
        return migrated;
    }

    return tomlToDatabasesConfig(allocator, toml);
}

//
// Resolves the legacy recent-paths array into the new recent-names array by looking up
// each path in the current databases list. Paths that don't match any entry are dropped.
//
fn recentPathsToNames(allocator: std.mem.Allocator, recentPaths: []const []const u8, databases: []const IDatabaseEntry) ![]const []const u8 {
    var result: std.ArrayList([]const u8) = .empty;
    for (recentPaths) |recentPath| {
        for (databases) |dbEntry| {
            if (std.mem.eql(u8, dbEntry.path, recentPath)) {
                try result.append(allocator, dbEntry.name);
                break;
            }
        }
    }
    return result.items;
}

//
// Saves the databases configuration to disk.
// (Zig: the TypeScript coercion of missing arrays to [] is not needed, because the slices always exist.)
//
pub fn saveDatabasesConfig(allocator: std.mem.Allocator, io: std.Io, config: IDatabasesConfig) !void {
    try fs.writeToml(allocator, io, try DATABASES_FILE(allocator), try databasesConfigToToml(allocator, config));
}

//
// Returns all configured database entries.
//
pub fn getDatabases(allocator: std.mem.Allocator, io: std.Io) ![]const IDatabaseEntry {
    const config = try loadDatabasesConfig(allocator, io);
    return config.databases;
}

// Not ported: findDatabase, addDatabaseEntry, updateDatabaseEntry, removeDatabaseEntry, getRecentDatabases,
// removeRecentDatabaseName, markDatabaseOpened (psi dbs, not psi replicate or psi verify).
